Mempool Live Statistics Panel

priyashuu opened the pull request:
###  Overview

This PR introduces a **comprehensive Mempool monitoring panel** that displays real-time transaction data, fee estimations, and block fee history. It enhances the visibility of mempool activity by consuming live data via WebSocket from `server.js`, which can be launched via:

```bash
node dashboard/api/server.js
```

---

### What's Included


####  Realtime Data Visualization

* **Fee Rate Distribution**:
  * Bar chart showing current fee rate percentiles (min, median, 90th, max).
* **Block Fee History**:

  * Line chart of latest blocks showing:

    * BTC fees
    * USD fees
  * Toggle between `BTC`, `USD`, or `both` views.

####  Stat Cards Display

Real-time statistics via dynamic components:

*  Mempool Size (vB)
* Transaction Count
*  Total Fees (BTC | USD)
*  Next Block Fee Preview
*  Fee Estimates by Priority:

  * High / Medium / Low / No Priority

---

### How to Testify

1. Create a .env file at the root of the api directory and add the required environment variables:
2. Run the WebSocket server:

```bash
node dashboard/api/server.js
```

3. Navigate to the **Mempool** tab in the UI.

4. Watch live updates flow into:

   * Top-level mempool stats
   * Fee estimate cards
   * Fee rate distribution histogram
   * Block fee history chart



### 📷 Preview

Here’s what the final interface looks like:
<img width="1892" height="896" alt="image" src="https://github.com/user-attachments/assets/d21c8a3e-7e10-4c17-ac6c-8d3ee0b369fa" />
<img width="1758" height="646" alt="image" src="https://github.com/user-attachments/assets/2d3044bb-8c70-4b12-81d3-37a817459120" />



Open to suggestions for improvements or enhancements!

mcelrath committed:
doc: add contribution guidelines (#156) (#157)

Co-authored-by: Mohd Zaid <72139621+zaidmstrr@users.noreply.github.com>

u32luke committed:
Update general_considerations.md (#173)

Typo fix

priyashuu committed:
Merge remote-tracking branch 'origin/dev' into update-mempool

priyashuu committed:
feat: integrate mempool stats with WebSocket broadcast

priyashuu committed:
add BTC/USD toggle and custom styled tooltip

priyashuu committed:
Merge remote-tracking branch 'origin/dev' into update-mempool

priyashuu committed:
Started with a single data point to keep things consistent with how other graphs are initialized. Removed 7 weeks of historical data for now

priyashuu committed:
refactor(core): introduce typed interfaces for Fee and Mempool data

- Replaced all 'any' types with strict interfaces (Fee, FeeEstimate, MempoolData, etc.)
- Enhanced WebSocket condition checks to handle missing or malformed data gracefully

priyashuu committed:
Merge remote-tracking branch 'origin/dev' into update-mempool

priyashuu committed:
resolved the conflicts and prettier formatting

its-me-abhishek commented on dashboard/src/components/Dashboard/Dashboard.tsx:
```diff
@@ -3,7 +3,7 @@ import Card from '../common/Card';
 import InstallationInstructions from '../Installation/InstallationInstructions';
 import MineInventoryDashboard from '../MinerInventory/MineInventoryDashboard';
-import MempoolLatencyStats from './MempoolLatencyStats';
+import Mempool from '../Mempool/Mempool';
```
please revert this back, & change file name back if possible ; as changing names usually leads to confusion

its-me-abhishek commented on dashboard/src/components/Mempool/Mempool.tsx:
```diff
@@ -0,0 +1,291 @@
+import { useEffect, useState, useRef } from 'react';
```
maybe you forgot to add./update tests

its-me-abhishek reviewed (changes requested):
LGTM, just minor changes required

priyashuu committed:
test: add MempoolLatencyStats.test.tsx and restore original filename

priyashuu committed:
Merge remote-tracking branch 'origin/dev' into update-mempool

priyashuu committed:
Merge remote-tracking branch 'origin/dev' into update-mempool

priyashuu committed:
test: add comprehensive tests for fetchMempoolStats function

priyashuu committed:
Merge remote-tracking branch 'origin/dev' into update-mempool

its-me-abhishek commented on dashboard/api/utils/fetchMempoolStats.js:
```diff
@@ -0,0 +1,91 @@
+  try {
+    const statsRes = await axios.get(`${process.env.MEMPOOL_STATS_URL}`);
+    const feesRes = await axios.get(`${process.env.FEE_RECOMMENDED_URL}`);
+    const priceRes = await axios.get(`${process.env.BTC_PRICE_URL}`);
```
Please add some reference/Docs/Readme/comments for these URLs

its-me-abhishek commented on dashboard/src/components/Mempool/MempoolLatencyStats.tsx:
```diff
@@ -0,0 +1,291 @@
+                name="BTC"
+              />
+            )}
+            {(selectedView === 'usd' || selectedView === 'both') && (
```
Can we have a currency selector here? similar to BTC stats page? For better insights.

its-me-abhishek reviewed (commented):
Please review, had some suggestions

priyashuu committed:
refactor: update environment variable names and improve fetchMempoolStats function URLs

priyashuu committed:
fixes

priyashuu committed:
added jpy ,andeur currecy in the block fee chart , and updated the ltests file also

priyashuu committed:
remove fallback values for currency

priyashuu committed:
Fix: correct block fee currency conversion for EUR and JPY

- Removed unnecessary division by 100 for EUR and JPY values in block fee history
- USD still divided by 100 to convert from cents to dollars

mcelrath committed:
Merge branch 'dev' into update-mempool

mcelrath committed:
Make .env.example more verbose, remove broken binance url

    * The Binance API blocks some users and not others, for some reason

priyashuu committed:
fixed failed tests ,and removed the broken binace url

priyashuu committed:
I’ve updated the .env.example by removing MINER_DEVICE_URL, since we no longer need it as an environment variable.removed the console.logs for better productions code

priyashuu committed:
Merge remote-tracking branch 'origin/dev' into update-mempool

priyashuu committed:
resolved merge conflicts

priyashuu committed:
fix(jest): add moduleNameMapper for @ alias

- Updated jest.config.js to map `@/` to `src/` so Jest can resolve imports like `@/URLs`
- Fixes failing test in MempoolLatencyStats due to unresolved module path

mcelrath reviewed (approved):
Looks good!
